### ISSUE TITLE
Got different result with 2 perdict ways(seg task)

### ISSUE THREAD

**Issue by PengPeng-JunJun** [ultralytics/ultralytics#23609]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I perdict same image by 2 ways. After export onnx, I use yours code perdict and get the same result like by pt file.But when I do it with opencv DNN moduel,the mask pixel all zero.Can you solve my question? This is my opencv code

# 加载 YOLOv8 ONNX 分割模型
net = cv2.dnn.readNetFromONNX('E:\\py_Project\\ultralytics\\ultralytics\\yolo\\v8\\segment\\runs\\segment\\train2\\weights\\best.onnx')

# 加载输入图像
image = cv2.imread('E:\\py_Project\\Segmentation\\ENet\\LED2\\20251225115004-554.png')
height, width = image.shape[:2]

# 预处理输入图像
blob = cv2.dnn.blobFromImage(image, 1/255.0, (640, 640), (0, 0, 0), swapRB=True, crop=False)

# 设置网络输入
net.setInput(blob)

# 前向推理
outs = net.forward()

# 假设 YOLOv8 分割模型输出的是多通道掩码
segmentation_mask = outs[0]  # 提取分割掩码（batch_size=1 时是 outs[0]）

# 检查输出是否为空
if segmentation_mask is None:
    print("Error: Segmentation mask is None.")
else:
    print(f"Segmentation mask shape: {segmentation_mask.shape}")

# 如果模型输出多个类别的掩码（例如多通道），选择最大概率类别的掩码
segmentation_mask = np.argmax(segmentation_mask, axis=0)  # 选择最大概率类别作为掩码

# 确保掩码是 uint8 类型，并且值为 [0, 255]
segmentation_mask = (segmentation_mask * 255).astype(np.uint8)

# 检查掩码形状和数据类型
print(f"Resized segmentation mask shape: {segmentation_mask.shape}")
print(f"Segmentation mask dtype: {segmentation_mask.dtype}")

# 检查掩码尺寸是否与输入图像一致
if segmentation_mask.shape != (height, width):
    print("Resizing segmentation mask to match image size...")
    segmentation_mask = cv2.resize(segmentation_mask, (width, height), interpolation=cv2.INTER_LINEAR)

# 使用 bitwise_and 进行掩码应用
segmented_image = cv2.bitwise_and(image, image, mask=segmentation_mask)

# 显示结果
cv2.imshow("Original Image", image)
cv2.imshow("Segmentation Result", segmented_image)

# 等待键盘输入并关闭显示窗口
cv2.waitKey(0)
cv2.destroyAllWindows()

**Comment by UltralyticsAssistant** [comment]
👋 Hello @PengPeng-JunJun, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

Thanks for the detailed description and code snippet—ONNX + OpenCV DNN segmentation output differences can depend heavily on preprocessing, output tensor format, and postprocessing steps. 😊 This is an automated response, and an Ultralytics engineer will also assist soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. In your case, an MRE should include:1)The exact `ultralytics` version2)How the ONNX was exported (the exact export command/code)3)Your full OpenCV DNN inference script (including OpenCV version)4)The raw `outs` shapes/values summary (e.g., print shapes, min/max)5)A single input image (or a link to a small reproducible sample) and the expected vs actual result screenshots 📸

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Your OpenCV DNN code is treating the ONNX output as a per-pixel segmentation map, but Ultralytics segmentation ONNX exports **two tensors**: a detection tensor containing boxes/class scores **plus mask coefficients**, and a second tensor with **mask prototypes**; if you `argmax()` the detection tensor you’ll end up with nonsense (often all-zero after your later steps) because it’s not an `H×W` mask. See the docs section on [exported segmentation output tensors](https://docs.ultralytics.com/modes/export/#what-do-the-output-tensors-represent-in-exported-yolo-models) and the exact mask construction logic in [`SegmentationPredictor.construct_result()`](https://docs.ultralytics.com/reference/models/yolo/segment/predict/) (it uses `proto` + coefficients, then scales/crops).

```python
out_names = net.getUnconnectedOutLayersNames()
outs = net.forward(out_names)          # expect 2 outputs for seg models
pred, proto = outs[0], outs[1]
print(pred.shape, proto.shape)         # pred: (1, 4+nc+nm, N), proto: (1, nm, ph, pw)
```

If you still want OpenCV DNN inference, you’ll need to replicate Ultralytics postprocessing (transpose `pred`, run NMS, then generate instance masks from `proto` and each detection’s mask coefficients, then scale to `orig_img`); if you share your `yolo checks` output plus the printed `pred/proto` shapes (as text), we can sanity-check you’re reading the outputs in the right order.
